File: Difficulty: Medium/Longest subarray with Atmost two distinct integers/longest-subarray-with-atmost-two-distinct-integers.cpp
class Solution {
  public:
    int totalElements(vector<int> &arr) {
        int n = arr.size();
        unordered_map<int, int> map;
        int start = 0, maxLen = 0;
        
        for(int end=0; end<n; end++) {
            map[arr[end]]++;
            while(map.size() > 2 && start < n) {
                map[arr[start]]--;
                
                if(map[arr[start]] == 0)
                    map.erase(arr[start]);
                    
                start++;
            }
            maxLen = max(maxLen, end - start + 1);
        }
        
        return maxLen;
    }
};

File: Difficulty: Medium/Longest subarray with Atmost two distinct integers/longest-subarray-with-atmost-two-distinct-integers.java
class Solution {
    public int totalElements(int[] arr) {
        HashMap<Integer, Integer> map = new HashMap<>();
        int left = 0, maxLen = 0;

        for (int right = 0; right < arr.length; right++) {
            map.put(arr[right], map.getOrDefault(arr[right], 0) + 1);

            while (map.size() > 2) {
                map.put(arr[left], map.get(arr[left]) - 1);
                if (map.get(arr[left]) == 0) {
                    map.remove(arr[left]);
                }
                left++;
            }

            maxLen = Math.max(maxLen, right - left + 1);
        }
        return maxLen;
    }
}